Framework/Geometry/inc/MantidGeometry/Instrument/ComponentVisitor.h

```diff
@@ -0,0 +1,51 @@
+#ifndef MANTID_GEOMETRY_COMPONENTVISITOR_H
+#define MANTID_GEOMETRY_COMPONENTVISITOR_H
+
+#include <vector>
+#include <cstddef>
+namespace Mantid {
+namespace Geometry {
+
+class ICompAssembly;
+class IDetector;
+class IComponent;
+
+/** ComponentVisitor : Visitor for components. Enables parsing of a full doubly
+  linked InstrumentTree without need for dynamic casts.
+
+  Copyright &copy; 2017 ISIS Rutherford Appleton Laboratory, NScD Oak Ridge
+  National Laboratory & European Spallation Source
+
+  This file is part of Mantid.
+
+  Mantid is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 3 of the License, or
+  (at your option) any later version.
+
+  Mantid is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+  File change history is stored at: <https://github.com/mantidproject/mantid>
+  Code Documentation is available at: <http://doxygen.mantidproject.org>
+*/
+class ComponentVisitor {
+public:
+  virtual void
+  registerComponentAssembly(const ICompAssembly &assembly,
+                            std::vector<size_t> &detectorIndexes) = 0;
```

Comment:
I actually implemented as a load of dynamic_casts originally. I then refactored it to this. I put in some timings before I did the refactoring and found it to be more than x2 faster doing it this way. That was for a relatively small instrument, for larger instruments it would probably make an even bigger difference.